gctree/branching_processes.py

```diff
@@ -987,3 +964,230 @@ def f(x):
+        cm_list[0] = (1, 1)
+    # extract vector of function values and gradient components
+    logf_data = [CollapsedTree._ll_genotype(c, m, p, q) for c, m in cm_list]
+    logf = np.array([[x[0] * count_ls[i]] for i, x in enumerate(logf_data)]).sum()
+    grad_ll_genotype = np.array(
+        [x[1] * count_ls[i] for i, x in enumerate(logf_data)]
+    ).sum(axis=0)
+    return logf, grad_ll_genotype
+
+
+def llforest(
+    cm_countlist,
+    p: np.float64,
+    q: np.float64,
+    marginal: bool = False,
+) -> Tuple[np.float64, np.ndarray]:
+    r"""Log likelihood of branching process parameters :math:`(p, q)` given `cm_countlist`, an iterable
+    containing tuples `(cm_counts, mult)`, where `cm_counts` is an iterable describing a tree, and `mult`
+    is the number of trees in the forest which can be described with the iterable `cm_counts`.
+
+    `cm_counts` is in the format expected as an argument to :meth:`lltree`.
+
+    If ``marginal=False`` (the default), compute the joint log likelihood
+    .. math::
+        \ell(p, q; T, A) = \sum_{i=1}^n\log\mathbb{P}(T_i, A_i \mid p, q),
+    otherwise compute the marginal log likelihood
+    .. math::
+        \ell(p, q; T, A) = \log\left(\sum_{i=1}^n\mathbb{P}(T_i, A_i \mid p, q)\right).
+    Args:
+        p: branching probability
+        q: mutation probability
+        marginal: compute the marginal likelihood over trees, otherwise compute the joint likelihood of trees
+    Returns:
+        Log likelihood :math:`\ell(p, q; T, A)` and its gradient :math:`\nabla\ell(p, q; T, A)`
+    """
+    terms = [[lltree(cmcounts, p, q), count] for cmcounts, count in cm_countlist]
+    ls = np.array([term[0][0] for term in terms])
+    grad_ls = np.array([term[0][1] for term in terms])
+    # This can be done ahead of time
+    count_ls = np.array([term[1] for term in terms])
+    if marginal:
+        # we need to find the smallest derivative component for each
+        # coordinate, then subtract off to get positive things to logsumexp
+        grad_l = []
+        for j in range(len((p, q))):
+            i_prime = grad_ls[:, j].argmin()
+            b = (grad_ls[:, j] - grad_ls[i_prime, j]) * count_ls
+            # believe it or not, logsumexp can't handle 0 in b
+            # when np.seterr(underflow='raise') on newer processors:
+            grad_l.append(
+                grad_ls[i_prime, j]
+                + np.exp(
+                    scs.logsumexp((ls - ls[i_prime])[b != 0], b=b[b != 0])
+                    - scs.logsumexp(ls - ls[i_prime], b=count_ls)
+                )
+            )
+        # count_ls shouldn't have any zeros in it...
+        return (-np.log(count_ls.sum()) + scs.logsumexp(ls, b=count_ls)), np.array(
+            grad_l
+        )
+    else:
+        return (ls * count_ls).sum(), np.array(
+            [(grad_ls[:, 0] * count_ls).sum(), (grad_ls[:, 1] * count_ls).sum()]
+        )
+
+
+def fit_branching_process(dag, verbose=True, marginal=True):
```

Comment:
Does this need to be a public function in the API that is shown in docs? If not, suggest renaming `_fit_branching_process `.

Reply:
I don't know, I had kind of thought that the inference pipeline should be reproducible using functions in the API. Like if you don't want to use the cli, all the pieces that it puts together are accessible and documented. I have no idea if that's a goal, or what sorts of things you want in the API, in general.

Reply:
Ok I agree with that. But note that this function isn't currently documented, because the `branching_processes` module isn't documented. The tree and forest classes in that module are promoted to package-level classes (see `gctree/__init__.py`), and those are documented. We'd need to add `branching_processes` in Modules section of `api.rst` (and probably a note in the Classes section explaining that these are promoted from the `branching_processes` module).

Reply:
Also note missing type hint on `dag` argument.

Reply:
I think this is all taken care of, we'll see when I build the docs again.